### DISCUSSION TITLE
Training memory usage

### DISCUSSION THREAD

**Discussion by jcnewell** [explosion/spaCy#7132] [Help: Coding & Implementations] [answered]
Hi,
Training a textcat appears to require substantially more memory than spacy2. _spacy train_ is using 44 GB of memory, compared to less than 4 GB previously using the Quickstart config for both _efficiency_ and _accuracy_. Are there any settings which might reduce this?

**Comment by svlandeg** [comment, marked as answer]
Does your `textcat` problem have mutually exclusive labels, or can there be multiple correct labels per instance? And can you share the config file you're using with `spacy train`? And perhaps also the code you used when you were training it with spaCy v2? 

In general we tried to keep the "default" parameters pretty much the same between the "old way" in v2, and the new default config files in v3, assuming that you're using a `tok2vec` and not a `transformer`. But I can imagine that perhaps some setting is still different, like the batch size, or width of the embedding layer, etc. But it's difficult to say without much more information ;-)

**Reply by jcnewell** [reply]
I'm using the default config provided by the Quickstart utility for:

    Language: English
    Components: textcat
    Hardware: CPU
    Optimisation: efficiency

With these defaults exclusive_classes is true. The spacy2 version used the defaults and approach in:
https://github.com/explosion/spaCy/blob/v2.x/examples/training/train_textcat.py

The training data has a single binary label with 1.6M samples (max text words: 200, mean text words: 39)
BTW the spacy3 version also appears to only use one CPU.
I'll try adjusting some of the config parameters.

**Reply by svlandeg** [reply]
One significant difference that would have an impact on memory usage, is the `batch_size`. In the v2 script you linked, they are defined by 
```
batch_sizes = compounding(4.0, 32.0, 1.001)
```
but the v3 config has by default:
```
[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0
```
so they'll be much larger overall. What happens if you adjust the `start` and `stop` to 4 and 32 respectively?

**Reply by jcnewell** [reply]
Strangely, this has no effect on the memory usage. Starts around 4 GB but rises to 44 GB after the headers for the evaluation scores appear. I'm going to adapt the spacy2 script for spacy3 and see how that goes.

**Reply by jcnewell** [reply]
FYI a spacy3 version of the spacy2 training script behaves the same as using _spacy train_ e.g. 44 GB memory and 100% CPU
For comparison, using the original spacy2 training script uses 1.68 GB and 780% CPU

**Reply by adrianeboyd** [reply]
The current training loop has the whole corpus in memory after the first epoch for easy shuffling in future epochs, which is probably what accounts for this memory usage (1.6 M * 2 `Doc` objects). The `Example` objects are larger than their counterparts in v2, especially for textcat, and the simple training loop used to do the `GoldParse` conversion on-the-fly for each batch so it didn't end up having as many larger objects in memory at the same time.

We're thinking about how to modify the default training loop so it works better with very large or infinite training corpora. Then the default training loop can't do the shuffling so you'd need to move that into the reader. See: #7208
